File: packages/express/src/utils/route-registry.ts
export class RouteRegistry {
  private readonly routes = new Map<string, Set<string>>();
  private readonly overrides = new Set<string>();
  private readonly customRoutes = new Set<string>();

  register(path: string, methods: string[], meta?: { entity?: string; relationship?: string; isCustom?: boolean }): void {
    const key = this.normalizePath(path);
    const existing = this.routes.get(key) ?? new Set<string>();
    for (const method of methods) {
      existing.add(method.toUpperCase());
    }
    this.routes.set(key, existing);

    if (meta?.isCustom) {
      this.customRoutes.add(key);
    }
  }

  markOverride(method: string, path: string): void {
    this.overrides.add(this.overrideKey(method, path));
  }

  isOverridden(method: string, path: string): boolean {
    return this.overrides.has(this.overrideKey(method, path));
  }

  summary(): Record<string, unknown> {
    return {
      routeCount: this.routes.size,
      overrideCount: this.overrides.size,
      customRouteCount: this.customRoutes.size,
      routes: Object.fromEntries(
        [...this.routes.entries()].map(([path, methods]) => [path, [...methods]]),
      ),
    };
  }

  private overrideKey(method: string, path: string): string {
    return `${method.toUpperCase()}:${this.normalizePath(path)}`;
  }

  private normalizePath(path: string): string {
    return path.replace(/:([A-Za-z0-9_]+)/g, '{$1}');
  }
}

File: packages/fastapi/fastbackend_fastapi/utils/route_registry.py
"""Route registry for dynamic route management."""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class RegisteredRoute:
    path: str
    methods: list[str]
    entity: str | None = None
    relationship: str | None = None
    is_custom: bool = False
    is_override: bool = False


@dataclass
class RouteRegistry:
    routes: list[RegisteredRoute] = field(default_factory=list)
    overrides: set[str] = field(default_factory=set)

    def register(
        self,
        path: str,
        methods: list[str],
        *,
        entity: str | None = None,
        relationship: str | None = None,
        is_custom: bool = False,
        is_override: bool = False,
    ) -> None:
        self.routes.append(
            RegisteredRoute(
                path=path,
                methods=methods,
                entity=entity,
                relationship=relationship,
                is_custom=is_custom,
                is_override=is_override,
            )
        )

    def mark_override(self, method: str, path: str) -> None:
        self.overrides.add(f"{method.upper()}:{path}")

    def is_overridden(self, method: str, path: str) -> bool:
        return f"{method.upper()}:{path}" in self.overrides

    def summary(self) -> dict[str, int]:
        return {
            "total": len(self.routes),
            "crud": sum(1 for r in self.routes if r.entity and not r.is_custom),
            "relationships": sum(1 for r in self.routes if r.relationship),
            "custom": sum(1 for r in self.routes if r.is_custom),
            "overrides": len(self.overrides),
        }
